lib/SILOptimizer/Mandatory/DataflowDiagnostics.cpp

```diff
@@ -46,17 +46,23 @@ static void diagnoseMissingReturn(const UnreachableInst *UI,
     llvm_unreachable("unhandled case in MissingReturn");
   }
 
-  // No action required if the function returns 'Void' or that the
-  // function is marked 'noreturn'.
-  if (ResTy->isVoid() || F->isNoReturnFunction())
+  // No action required if the function returns 'Void'.
+  if (ResTy->isVoid())
     return;
 
   SILLocation L = UI->getLoc();
   assert(L && ResTy);
-  diagnose(Context,
-           L.getEndSourceLoc(),
-           diag::missing_return, ResTy,
-           FLoc.isASTNode<ClosureExpr>() ? 1 : 0);
+  if (F->isNoReturnFunction()) {
+    diagnose(Context,
+             L.getEndSourceLoc(),
+             diag::missing_never_call,
```

Comment:
You can combine both diagnose() calls into one and use the ternary operator to select the diag